File: decode_hex.py
from eth_utils import decode_hex
from eth_abi.abi import decode

def decode_hex_data(hex_data):
    """Decode hex-encoded data for the staking contract."""
    try:
        # Remove the 0x prefix if it exists
        hex_data = hex_data[2:] if hex_data.startswith("0x") else hex_data

        # Function selector is the first 4 bytes, skip it (8 characters in hex)
        params = decode(['uint256', 'uint256'], decode_hex(hex_data[8:]))

        # Parse the decoded data
        validator_id = str(params[0])  # First parameter: Validator ID
        amount_in_tokens = params[1] / 10**18  # Convert from Wei to tokens

        return {
            "validatorId": validator_id,
            "amountInTokens": str(amount_in_tokens)
        }
    except Exception as e:
        print(f"Error decoding hex data: {e}")
        return None

File: decodeHex.js
const BN = require('bn.js');

// Get the hex data from the command-line arguments
const hexData = process.argv[2];
if (!hexData) {
  console.error("Hex data argument is required.");
  process.exit(1);
}

/**
 * Decode hex data for the staking contract.
 * @param {string} hexData - The hex-encoded transaction data.
 * @returns {object} Decoded data including validatorId and token amount.
 */
function decodeHexData(hexData) {
  // Break the data into 32-byte chunks (skipping function selector)
  const params = hexData.slice(10).match(/.{1,64}/g);

  // Decode the parameters as 256-bit integers
  const decodedParams = params.map(param => new BN(param, 16));

  // Extract simplified fields
  const validatorId = decodedParams[0].toString(); // First parameter
  const amountInTokens = decodedParams[1].div(new BN('1000000000000000000')).toString(); // Convert to tokens

  return {
    validatorId,
    amountInTokens,
  };
}

// Decode and output the result
const result = decodeHexData(hexData);
console.log(JSON.stringify(result));
